libevmasm/PeepholeOptimiser.cpp

```diff
@@ -572,14 +572,31 @@ struct UnreachableCode
 		auto end = _state.items.end();
 		if (it == end)
 			return false;
-		if (
-			it[0] != Instruction::JUMP &&
-			it[0] != Instruction::RJUMP &&
-			it[0] != Instruction::RETURN &&
-			it[0] != Instruction::STOP &&
-			it[0] != Instruction::INVALID &&
-			it[0] != Instruction::SELFDESTRUCT &&
-			it[0] != Instruction::REVERT
+		if (it[0].type() == Operation)
+		{
+			// These instructions have their own AssemblyItemType
+			solAssert(
+				it[0] != Instruction::RJUMP &&
+				it[0] != Instruction::RETURNCONTRACT &&
+				it[0] != Instruction::JUMPF &&
+				it[0] != Instruction::RETF
+			);
+
+			if (
+				it[0] != Instruction::JUMP &&
+				it[0] != Instruction::RETURN &&
+				it[0] != Instruction::STOP &&
+				it[0] != Instruction::INVALID &&
+				it[0] != Instruction::SELFDESTRUCT &&
+				it[0] != Instruction::REVERT
+			)
+				return false;
+		}
+		else if (
+			it[0].type() != RelativeJump &&
+			it[0].type() != JumpF &&
+			it[0].type() != RetF &&
+			it[0].type() != ReturnContract
 		)
```

Comment:
The reason I wanted these new items to store the instruction was exactly so that we can treat them uniformly in situations like this. You can pretty much disregard the type and only look at the instruction.

```suggestion
		if (
			it[0] != Instruction::JUMP &&
			it[0] != Instruction::RJUMP &&
			it[0] != Instruction::RETURNCONTRACT &&
			it[0] != Instruction::JUMPF &&
			it[0] != Instruction::RETF &&
			it[0] != Instruction::RETURN &&
			it[0] != Instruction::STOP &&
			it[0] != Instruction::INVALID &&
			it[0] != Instruction::SELFDESTRUCT &&
			it[0] != Instruction::REVERT
		)
			return false;
```

If you want to be 100% sure we won't miss some weird item where someone forgot to set the instruction, you could also add an assert in this `if` that if we're skipping it, it's not one of those 4 types. But I'd say it's optional.

Reply:
Sure. This code was added to PoC branch before the first EOF branch was merged. Forgot about it.

Reply:
Unfortunately I cannot use `==` operator to compare with `Instruction` b/c it checks that type equals `Operation`